buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesTask.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.gradle.plugin;
+
+import org.elasticsearch.gradle.Version;
+import org.elasticsearch.gradle.VersionProperties;
+import org.gradle.api.InvalidUserDataException;
+import org.gradle.api.Project;
+import org.gradle.api.Task;
+import org.gradle.api.tasks.Copy;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.InputFile;
+import org.gradle.api.tasks.OutputFile;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.StandardCopyOption;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Creates a plugin descriptor.
+ */
+public class PluginPropertiesTask extends Copy {
+
+    private PluginPropertiesExtension extension;
+
+    private File descriptorOutput = new File(getProject().getBuildDir(),
+            "generated-resources/plugin-descriptor.properties");
+
+    public PluginPropertiesTask() {
+        Project project = this.getProject();
+
+        File templateFile = new File(project.getBuildDir(), "templates/" + this.descriptorOutput.getName());
+
+        Task copyPluginPropertiesTemplate = project.getTasks().create("copyPluginPropertiesTemplate")
```

Comment:
We don't want to configure in the constructor for anything Java related. 
Can you please rewrite this to use `project.copy` in the `@TaskAction`.

The validations on the `afterEvaluate` should go into `getExtension` since we want these to fail even if the task actions ( i.e. the task itself ) are not executed.

Reply:
@atorok I would need some feedback around what I have done for 
```
We don't want to configure in the constructor for anything Java related.
Can you please rewrite this to use project.copy in the @TaskAction.
``` 
I wrote something as follow inside `@TaskAction` 
```
project.copy((unused) -> {
       from("/" + this.descriptorOutput.getName());
       into(templateFile.toPath());
 });
```
but, what about the line with `dependsOn`? We don't have a reference to task. First, time I am working with `Gradle` hence, why the struggle I guess.

Reply:
You won't need the `dependsOn` ( it's `this.dependsOn` btw ). 
And you need to ditch the `extends Copy` just  extend `DefaultTask`. 
`project.copy` behaves like the `Copy` task except it runs when you call it, it's not a task, but it's configured in the same way. The `unused` variable is what you use to configure it. What you wrote above is a nop, a copy that does nothing but configure the task it's called in as  then  `from` refers to configuration on the task, that has no effect anyway because we are already in the execution phase when `@TaskAction` runs.  Since you no longer extend the `Copy` task, you'll have to use another `project.copy` for that functionality.  There will be no `project.tasks.create('copyPluginPropertiesTemplate')` either, as that's going to be replaced with  a `project.copy` too.

Reply:
I pushed some changes along with unit tests 👍